Microsoft.Toolkit.Mvvm/ComponentModel/ObservableRecipient.cs

```diff
@@ -0,0 +1,196 @@
+{
+    /// <summary>
+    /// A base class for observable objects that also acts as recipients for messages. This class is an extension of
+    /// <see cref="ObservableObject"/> which also provides built-in support to use the <see cref="IMessenger"/> type.
+    /// </summary>
+    public abstract class ObservableRecipient : ObservableObject
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableRecipient"/> class.
+        /// </summary>
+        /// <remarks>
+        /// This constructor will produce an instance that will use the <see cref="Messaging.Messenger.Default"/> instance
+        /// to perform requested operations. It will also be available locally through the <see cref="Messenger"/> property.
+        /// </remarks>
+        protected ObservableRecipient()
+            : this(Messaging.Messenger.Default)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ObservableRecipient"/> class.
+        /// </summary>
+        /// <param name="messenger">The <see cref="IMessenger"/> instance to use to send messages.</param>
+        protected ObservableRecipient(IMessenger messenger)
+        {
+            Messenger = messenger;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="IMessenger"/> instance in use.
+        /// </summary>
+        protected IMessenger Messenger { get; }
+
+        private bool isActive;
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the current view model is currently active.
+        /// </summary>
+        public bool IsActive
+        {
+            get => this.isActive;
+            set
+            {
+                if (Set(ref this.isActive, value, true))
+                {
+                    if (value)
+                    {
+                        OnActivated();
+                    }
+                    else
+                    {
+                        OnDeactivated();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Raised whenever the <see cref="IsActive"/> property is set to <see langword="true"/>.
+        /// Use this method to register to messages and do other initialization for this instance.
+        /// </summary>
+        /// <remarks>
+        /// The base implementation registers all messages for this recipients that have been declared
+        /// explicitly through the <see cref="IRecipient{TMessage}"/> interface, using the default channel.
+        /// For more details on how this works, see the <see cref="MessengerExtensions.Register"/> method.
+        /// If you need more fine tuned control, want to register messages individually or just prefer
+        /// the lambda-style syntax for message registration, override this method and register manually.
+        /// </remarks>
+        protected virtual void OnActivated()
+        {
+            Messenger.Register(this);
```

Comment:
What does this (`Messenger.Register(this);`) do?
It registers the ObservableRecipient to receive messages, but there's no handler for receiving a message. I assume it's not redundant.
I may be missing something obvious, and it may just be a documentation issue, but: people don't read documentation, and if I don't understand it, then someone else may struggle too.

Reply:
This is the bit that was suggested by @michael-hawker, it basically automatically registers all the handlers declared in the class (if any). Specifically, it invokes this extension method using the current instance as recipient:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/49d2ea11c947621d8276a60cafcc8a2d1531afe7/Microsoft.Toolkit.Mvvm/Messaging/MessengerExtensions.cs#L56-L57

What it means is, consider this simple example:

```csharp
public class MyViewModel : ObservableRecipient, IRecipient<Message>, IRecipient<MessageB>
{
    public void Receive(MessageA message) { }

    public void Receive(MessageB message) { }
}
```

In this case, `MyViewModel` has two declared receivers (using the `IRecipient<TMessage>` interface). When a dev sets the `IsActive` property for the viewmodel to `true`, that `OnActivated` method will be called and it'll automatically registers those two messages for the receiver (the viewmodel) using the current `IMessenger` instance (if there is no message handlers declared, that method will simply do nothing). Similarly, `OnDeactivated` will automatically unregister all the existing message handlers. This is done to make the code less verbose, as all message handlers are automatically added. Of course devs are free to just override either (or both) of those methods and setup their handlers manually (eg. using other APIs that don't require the `IRecipient<TMessage>` interface). One big point here was offering flexibility to choose the preferred approach to configure things up.

I'll write this here but this is meant as a reply to your other two comments as well, yeah we're absolutely working on the docs, and I do plan to add extensive info about especially these two base classes. I think I'll also add some sort of FAQ section with small questions and code snippets to show the various approaches etc. and I'll also expand the XML comments right above each class/method, to ensure even devs that don't go read the docs will be able to just read all the necessary info right from the IntelliSense popup in Visual Studio (or their editor of choice, or the API browser, or something else) 😊

Reply:
Ok, I'm sold on getting rid of the `ViewModelBase` name and now I understand how the messenger registration works but I'm not convinced on the `Register` method names. 
The overloads (with and without a receiver being passed) are doing very different things. Giving them different names to show they do different things could help with understanding. (perhaps `RegisterAllMessages()` without passing a recipient and `RegisterMessage()` when one is passed)

Reply:
> Ok, I'm sold on getting rid of the `ViewModelBase` name 

Ahahah awesome, glad to hear that! 🥳
Especially because I take it that means the reasoning behind the change in general makes sense to others as well, perfect then!

> The overloads (with and without a receiver being passed)

I'm confused about this. All the `Subscribe` overloads _always_ take the target recipient. They can take different parameters after that (if any), but the recipient itself is always present and it's always the first parameter 🤔

Reply:
> > The overloads (with and without a receiver being passed)
> 
> I'm confused about this. All the `Subscribe` overloads _always_ take the target recipient. They can take different parameters after that (if any), but the recipient itself is always present and it's always the first parameter 🤔

I know that the recipient is passed in both but the receiver/action isn't.

Current:

```cs
// Pass a recipient but no receiver
Messenger.Register(this);

// pass a recipient and a receiver
Messenger.Register<SomethingHappenedMessage>(this, m => { ...  });
```

Disambiguated alternative:

```cs
// Pass a recipient but no receiver
Messenger.RegisterAllReceivers(this);

// pass a recipient and a receiver
Messenger.RegisterMessageHandler<SomethingHappenedMessage>(this, m => { ...  });
```

This is a space with lots of similar names and loaded terms.
You've just mentioned "Subscribe" and this is another common term in the messaging space but not used in this code.

As you're still using this language, maybe we could use this term to avoid overloading "Register". This also better reflects that this is a version of the pub/sub pattern.

`Messenger.Subscribe<SomethingHappenedMessage>(this, m => { ...  });`

While having overloads can be useful it also risks making it difficult to understand which to use, or what the subtle differences between the version are.

---

My primary concern was that `Messenger.Register(this);` raises a lot of questions. 
- If `Messenger` is a property of the ObservableCollection, why does it need a reference to "this"? That looks like a weird coupling.
- What is being registered?
- How can this register anything if there are no messages configured?
- Will this register/subscribe to all messages if none are defined? If none, why do this?
- How will messages be received?

Yes, we've answered these but I hope a different name could avoid others from asking them.

Reply:
> I know that the recipient is passed in both but the receiver/action isn't.

Ooh I see where my confusion was coming from here, we're using the same word for different concepts 😄

In general, I consider recipient, subscriber and receiver to be synonyms in this context, whereas the target `Action<TMessage>` delegate would be the message handler (or just the action). So I was struggling to follow here as you were referring to the action as "receiver", while for me the receiver would've just been the target recipient/subscriber.

> Disambiguated alternative:
> 
> ```cs
> // Pass a recipient but no receiver
> Messenger.RegisterAllReceivers(this);
> 
> // pass a recipient and a receiver
> Messenger.RegisterMessageHandler<SomethingHappenedMessage>(this, m => { ...  });
> ```

I have to say I'm not really convinced by this, especially because this would make the code much more verbose. Eg. users not using the `IRecipient<TMessage>` interface and instead using the `MvvmLight`-like pattern, manually registering to each message, would then have a much longer name for every single message registration. Considering that this is often done in one-liners, this would be an issue. For instance, consider this snippet from one of my apps:

```csharp
Messenger.Register<MemoryStateRequestMessage>(this, m => m.Reply(MachineState));
Messenger.Register<RunCommandRequestMessage>(this, m => _ = ExecuteCommandAsync());
Messenger.Register<DeleteOperatorRequestMessage>(this, m => _ = DeleteLastOperatorAsync());
Messenger.Register<ClearCommandRequestMessage>(this, m => _ = ResetCommandAsync());
Messenger.Register<RestartConsoleRequestMessage>(this, m => _ = RestartAsync());
Messenger.Register<ClearConsoleScreenRequestMessage>(this, m => _ = ClearScreenAsync());
Messenger.Register<RepeatCommandRequestMessage>(this, m => _ = RepeatLastScriptAsync());
Messenger.Register<OverflowModeSettingChangedMessage>(this, m => _ = RestartAsync());
Messenger.Register<MemorySizeSettingChangedMessage>(this, m => _ = RestartAsync());
```

Keeping the `Register` name as short as possible is pretty important in these scenarios.
What I'd suggest instead would be to have two method names:

- `Register` for all the overloads taking an `Action<TMessage>`, so registering a _single_ handler
- `RegisterAll` for the overloads only taking the recipient, and then adding all handlers automatically

With this change, your previous sample would become:

```csharp
// Pass a recipient but no receiver (multiple/all handlers)
Messenger.RegisterAll(this);

// Pass a recipient and a receiver (single handler)
Messenger.Register<SomethingHappenedMessage>(this, m => { ...  });
```

Similarly, I guess we could also have `Unregister` and `UnregisterAll` depending on whether an overload is only looking for and removing a single handler, or all handlers for a given recipient (or for a recipient, on a specified channel).

What about this? 😊